Use single-target partybot buffs when it makes more sense.

mserajnik opened the pull request:
## 🍰 Pullrequest
<!-- Describe the Pullrequest. -->
This fixes Mage partybots not reapplying Arcane Brilliance unless it has expired on themselves.

It also makes it so that partybots now choose either the single target or party-wide variant of the following buffs:

- Arcane Intellect / Arcane Brilliance
- Power Word: Fortitude / Prayer of Fortitude
- Divine Spirit / Prayer of Spirit
- Shadow Protection / Prayer of Shadow Protection
- Mark of the Wild / Gift of the Wild

The selection process is:
- If there is only a single party member missing the respective buff, use the single target version
- If there is more than one party member missing the buff, use the party-wide version

Of course, this also depends on which spells the partybot actually knows, as well as the party members within buff range.

### Proof
<!-- Link resources as proof -->
- None

### Issues
<!-- Which Issues does this fix, which are related?
- fixes #XXX
- relates #XXX
-->
- Fixes #2400

### How2Test
<!-- Give a detailed description how to test your PR and confirm it is working as expected.
- Test1
- Test2
-->
- Summon a Mage partybot that knows both Arcane Intellect and Arcane Brilliance
- Wait for it to cast Arcane Brilliance, then remove the buff from yourself
- Observe the Mage partybot now casting Arcane Intellect on you (once it has enough Mana), because it itself still has the Arcane Brilliance buff (so only one party member needs it -> choose Arcane Intellect)

Video demonstration:


https://github.com/user-attachments/assets/5c22e079-1583-4c1d-adce-40f4425f4027



### Todo / Checklist
<!-- In case some parts are still missing, important notes, breaking changes and other notable items, list them here. -->
- [X] None

mserajnik committed:
Fix Mage partybots not reapplying Arcane Brilliance.

gajet5 commented:
https://github.com/vmangos/core/pull/2510/commits

mserajnik commented:
@gajet5 

Sorry, I missed your PR somehow; you should mark that it closes #2400.

mserajnik commented:
I'm reopening this because @gajet5 closed his PR (#2510) and this is a tested fix for #2400.

mserajnik commented:
No actual code changes in the force-push; I just updated this PR to be based on the latest `development` commit because when downloading the changes [as patch](https://github.com/vmangos/core/pull/2590.patch) it wasn't possible to `git apply` it anymore because the patch included the now renamed `UNIT_STAT_ROOT` in the surrounding lines of code:
![image](https://github.com/user-attachments/assets/d98e2bb7-6fc2-4abd-8666-562fbd384999)
(previous generated patch on the left, new patch after rebasing on the right)

mserajnik commented:
@ratkosrb Any chances to get this merged since you also dealt with other bot stuff in https://github.com/vmangos/core/commit/b530bcbd7a2359376d35be658661acf87be816f9? 😄

ratkosrb commented:
I don't think mages should cast Arcane Brilliance unless there are multiple targets missing the buff. For a single target they should cast Arcane Intellect.

mserajnik commented:
True, that makes more sense.

At the moment, other party-wide buffs like Prayer of Fortitude or Prayer of Spirit behave like that (= they get rebuffed once they drop off a single party member), only Arcane Brilliance behaves differently, so this PR would at least make it so the behavior is unified.

But I can adjust it so it works like you describe. I assume you would want the following spells behave the same?

- Power Word: Fortitude / Prayer of Fortitude
- Divine Spirit / Prayer of Spirit
- Shadow Protection / Prayer of Shadow Protection
- Mark of the Wild / Gift of the Wild

There is also the various Paladin Blessings [bots can cast](https://github.com/mserajnik/vmangos-core/blob/5ba1a6a90446c8b62262cc454cd2dc8bce9b86d9/src/game/PlayerBots/CombatBotBaseAI.cpp#L142-L146):

- Blessing of {Light, Might, Wisdom, Kings,, Sanctuary} / Greater Blessing of { ... }

But I guess we can disregard those since their are more specific (Greater Blessings will only apply to members of the same class as the target).

Did I miss anything else?

mserajnik committed:
Use single-target partybot buffs when it makes more sense.

Choose the single-target variant when exactly one eligible party member
is missing the buff, and only use the group variant when multiple
members need it.

This applies to:
- Arcane Intellect / Arcane Brilliance
- Power Word: Fortitude / Prayer of Fortitude
- Divine Spirit / Prayer of Spirit
- Shadow Protection / Prayer of Shadow Protection
- Mark of the Wild / Gift of the Wild

mserajnik committed:
Remove redundant guards from buff target helpers.

mserajnik commented:
I have adjusted the PR and marked it as Draft for now, since it's untested. Once I can confirm it's working as intended for me, I'll remove the Draft status.

mserajnik commented:
Tested locally and ready for review. A small video demonstration of a minimal test case I can think of (see __How2Test__ section in the PR message):

https://github.com/user-attachments/assets/0859e43c-774f-48ac-a0cc-7695fd22664a

@ShiyoKozuki: Maybe interesting for you if you are still using VMaNGOS.
